### PR TITLE
ksys573: Add incomplete support for Dance Dance Revolution Karaoke Mix

### PR DESCRIPTION
**Pull request by 987123879113**
The required karaoke I/O PCB is emulated just enough for it to be able to boot and inputs (except P2 side arrows) to work. Song data is not stored on the disc and an additional separate DAM karaoke system, which is unemulated, is required to actually play this.

Pic of PCB included for future reference. This is not my hardware so I don't have better pictures to provide. 
image

### CODE REVIEW

File: src/mame/machine/k573kara.cpp [hunk 1]
```diff
@@ -0,0 +1,172 @@
+// license:BSD-3-Clause
+// copyright-holders:windyfairy
+#include "emu.h"
+#include "k573kara.h"
+
+#define LOG_GENERAL    (1 << 0)
+// #define VERBOSE        (LOG_GENERAL)
+// #define LOG_OUTPUT_STREAM std::cout
+
+#include "logmacro.h"
```
Comment (cuavas) [inline src/mame/machine/k573kara.cpp:10]: This should come after all the other headers,

File: src/mame/machine/k573kara.cpp [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+	auto& duart_chan0(NS16550(config, "duart_com:chan0", clock() / 2));
+	auto& rs232_chan0(RS232_PORT(config, "rs232_chan0", default_rs232_devices, nullptr));
+	rs232_chan0.rxd_handler().set(duart_chan0, FUNC(ins8250_uart_device::rx_w));
+	rs232_chan0.dcd_handler().set(duart_chan0, FUNC(ins8250_uart_device::dcd_w));
+	rs232_chan0.dsr_handler().set(duart_chan0, FUNC(ins8250_uart_device::dsr_w));
+	rs232_chan0.ri_handler().set(duart_chan0, FUNC(ins8250_uart_device::ri_w));
+	rs232_chan0.cts_handler().set(duart_chan0, FUNC(ins8250_uart_device::cts_w));
+	duart_chan0.out_tx_callback().set(rs232_chan0, FUNC(rs232_port_device::write_txd));
+	duart_chan0.out_dtr_callback().set(rs232_chan0, FUNC(rs232_port_device::write_dtr));
+	duart_chan0.out_rts_callback().set(rs232_chan0, FUNC(rs232_port_device::write_rts));
```
Comment (cuavas) [inline src/mame/machine/k573kara.cpp:118]: If you bind things with references like this, bad things happen if some derived class uses `.replace()` later.

File: src/mame/machine/k573kara.cpp (hunk 1 above)
Comment (987123879113) [reply src/mame/machine/k573kara.cpp:10]: Looking at it now, I'm not using logging in this file anymore so I'll just remove it.

File: src/mame/machine/k573kara.cpp (hunk 2 above)
Comment (987123879113) [reply src/mame/machine/k573kara.cpp:112]: What would be the correct way to handle this? I looked for other similar code in `src/mame/machine` to see how other machines handled this case but almost everything seems to handle it the same way.

Comment (cuavas) [reply src/mame/machine/k573kara.cpp:112]: It’s safer to use tags if you don’t have a device finder for the thing, like:
```
 	rs232_chan0.rxd_handler().set("duart_com:chan0", FUNC(ins8250_uart_device::rx_w));
	rs232_chan0.dcd_handler().set("duart_com:chan0", FUNC(ins8250_uart_device::dcd_w));
	rs232_chan0.dsr_handler().set("duart_com:chan0", FUNC(ins8250_uart_device::dsr_w));
	rs232_chan0.ri_handler().set("duart_com:chan0", FUNC(ins8250_uart_device::ri_w));
	rs232_chan0.cts_handler().set("duart_com:chan0", FUNC(ins8250_uart_device::cts_w));
	duart_chan0.out_tx_callback().set("rs232_chan0", FUNC(rs232_port_device::write_txd));
	duart_chan0.out_dtr_callback().set("rs232_chan0", FUNC(rs232_port_device::write_dtr));
	duart_chan0.out_rts_callback().set("rs232_chan0", FUNC(rs232_port_device::write_rts)); 
```

Comment (987123879113) [reply src/mame/machine/k573kara.cpp:112]: Oh, I understand what you meant now. I'll make that change then. Thank you.

**Comment by cuavas** [comment]
Who should be credited for the dump?

**Pull request merged**